### DISCUSSION TITLE
set feature-state prior to first rendering?

### DISCUSSION THREAD

**Discussion by pdebruic** [maplibre/maplibre-gl-js#2374] [Q&A] [answered]
Hi - 

I've got 30,000 polygons and 400,000 points in mbtiles that I can render just fine.  I don't expect them to change much if at all ever, like once per year/quarter maybe.  

I have 350kb of json data that changes weekly I'd like to use to style them.  I could put the data into the tiles but then everyone would have to download MBs of new tiles weekly instead of just the data.  

Using feature-state I can conditionally change the color of the polygons and the points.  

Is there a way to set features state on the polygons/points from the source such that when each tile is loaded it uses the feature's state that's already been set?  

Right now I only know how to do it using `queryRenderedFeatures` or `querySourceFeatures` both of which, I think, give empty results until the tiles are loaded.  

Is there a way to pause or skip painting then resume it once the states have been updated?

Also - because `circle-sort-key` doesn't work with `feature-state` I've made 5 nearly identical layers to be able to vertically stack the points/polygons. Depending on what the user chooses to see points/polygons move up or down.   Is there a way to set a `z-index` on a feature?  

Thanks

**Comment by HarelM** [comment]
Do you have a small example you can share to better explain the problem? I'm not sure I fully understand it from what you wrote...

**Comment by pdebruic** [comment]
Yea sure lemme make one.  I'm making a dot density map that alternately shows isochrones, based on user choice.

Based on the variable in the json the user wants to look at,  I make quintiles and color the polygons and dots based on which quintile they're in.  Not all of the polygons or dots are shown for each variable.  

I want to sort them so the upper quintile dots and polygons are above the lower ones. Hence the multiple layers of identical data.

**Comment by pdebruic** [comment]
Here is a codesandbox that shows the gist of what I'm doing.  Its just for 4 polygons about 100 points.  When there is a ton of points or polygons profiling shows most of the time is spent in `queryRenderedFeatures`

https://codesandbox.io/s/crimson-frost-d676dv?file=/src/index.js

**Comment by pdebruic** [comment, marked as answer]
The answer to my problem was that I should have only been doing the `setFeatureState` called in response to a `data` event and only use the `querySourceFeatures` to find the features that needed it, and cacheing the ids of the ones I'd set the state for.  W
